### PR TITLE
Re-fetch shard info of primary when new node joins

### PR DESCRIPTION
**Pull request by dnhatn**
Today, we don't clear the shard info of the primary shard when a new node joins; then we might risk of making replica allocation decisions based on the stale information of the primary. The serious problem is that we can cancel the current recovery which is more advanced than the copy on the new node due to the old info we have from the primary.

With this change, we ensure the shard info from the primary is not older than any node when allocating replicas.

Relates #46959

This work was done by @henningandersen in #42518.
Co-authored-by: Henning Andersen <henning.andersen@elastic.co>

### CODE REVIEW

**Comment by dnhatn** [comment]
> Failure at [reference/cluster/health:36]: $body didn't match expected value:

This was fixed in #47016.

**Review by henningandersen — APPROVED** [review]
LGTM, but I am not sure my review counts on this one :smiley:...

File: server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            }
+            connection.sendRequest(requestId, action, request, options);
+        });
+        String newNode = internalCluster().startDataOnlyNode();
+        recoveryStarted.await();
+        // AllocationService only calls GatewayAllocator if there're unassigned shards
+        assertAcked(client().admin().indices().prepareCreate("dummy-index")
+            .setSettings(Settings.builder().put("index.routing.allocation.require.attribute", "not-found"))
+            .setWaitForActiveShards(0));
+        internalCluster().startDataOnlyNode(nodeWithReplicaSettings);
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java]: It looks to me like the recovery here always completes even without this change? (at least practically the test keeps passing when reverting the prod. code change for me).
I wasn't able to fully track down how that happens, but I figured I'd point it out :)

**Review by original-brownbear — COMMENTED** [review]
Just some drive-by comments

File: server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/gateway/ReplicaShardAllocatorIT.java]: I adjust the test in https://github.com/elastic/elasticsearch/pull/47035/commits/771582a4b591b20499cc8124d0f1f63166f4851a. We need to destroy the sync_id after the recovery on the new node has started. We clear the shard store cache when a shard is started. Hence, we fetch the shard store again when the-new-node joins. This PR improves a scenario when the content of the primary changes while a peer recovery is taking place.

**Review by original-brownbear — APPROVED** [review]
LGTM :) Thanks Nhat!
Might be best for David or Yannick to look over this as well though. I understand what's going on here just fine now, but might miss some implication of this change.

**Comment by dnhatn** [comment]
@DaveCTurner @ywelsch This PR blocks the work in https://github.com/elastic/elasticsearch/pull/46959. It would be great if one of you can take a look. Thank you!

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I suggested a comment and a few small changes.

**Review by DaveCTurner — APPROVED** [review]
LGTM thanks @dnhatn

**Comment by dnhatn** [comment]
Test failures are fixed in https://github.com/elastic/elasticsearch/issues/47196.

**Comment by dnhatn** [comment]
@henningandersen @original-brownbear @DaveCTurner Thank you for reviewing.

**Pull request merged**